Fix Python3/unicode issues

brianmay opened the pull request:
This started off being an issue in that cell2code calls encode()
on a str value, turning it into a bytes value, and then tries appending
it to a str value. Python doesn't like this.

Fixing this required trying to work out what the correct behaviour
was. Which required adding tests.

In writing the tests I found some issues that also affected python 2.7.
In particular, under Python 2 we need a lot more u prefixes and handling
of quotes in strings was broken.

Some of these problems are fixed, however there are still places that
are broken.

brianmay committed:
Fix Python3/unicode issues

This started off being an issue in that cell2code calls encode()
on a str value, turning it into a bytes value, and then tries appending
it to a str value. Python doesn't like this.

Fixing this required trying to work out what the correct behaviour
was. Which required adding tests.

In writing the tests I found some issues that also affected python 2.7.
In particular, under Python 2 we need a lot more u prefixes and handling
of quotes in strings was broken.

Some of these problems are fixed, however there are still places that
are broken.

brianmay commented:
Wondering how important is Python 2 support? I could make this patch simpler if dropping Python 2 support was considered OK.

Under Python 2, This patch changes everything using unicode strings. The alternative is that we could use UTF8 encoded str values. I think unicode is easier, but this could also potentially break client code???

brianmay commented on tests/ast/test_basic_evaluation.py:
```diff
@@ -71,6 +74,24 @@ def test_J2(self):
+        cell = self.sp.cellmap['Sheet1!J22']
+        assert cell.value == 'Paul'
+
+        assert self.sp.evaluate('Sheet1!J22') == 'Paul'
```
Noticed I have inconsistently used `assert` (for pytest) and `assertEqual` (legacy unittest standard). Is it OK if I change my `assertEqual` to `assert`?

brianmay replied:
Going to switch my additions to asserts, because they work far better with pytest which is used by travis and tox.

brianmay commented:
Forgot to mention, I updated `tests/ast/basic_evaluation.xlsx` to include a single unicode character. As I can't test unicode otherwise. That was the only change to the spreadsheet.

vallettea commented:
Thanks brian, however i’m on hollyday so i will only be able to push once i’m back.

Envoyé de mon iPhone

> Le 21 févr. 2019 à 07:12, Brian May <notifications@github.com> a écrit :
> 
> Forgot to mention, I updated tests/ast/basic_evaluation.xlsx to include a single unicode character. As I can't test unicode otherwise. That was the only change to the spreadsheet.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

brianmay committed:
Remove legacy testing stuff

This code is no longer required as we start tests with pytest and
was broken anyway.
